### PR TITLE
Changes required to run TiDB on Istio enabled k8s cluster

### PR DESCRIPTION
**Pull request by rahilsh**
### What problem does this PR solve?

Run TiDB on istio enabled k8s cluster
Closes: https://github.com/pingcap/tidb-operator/issues/4150

### What is changed and how does it work?
For TiDB to work on Istio enabled k8s cluster below are the configuration changes needed

1. Disable MTLS
```
---
apiVersion: "security.istio.io/v1beta1"
kind: "PeerAuthentication"
metadata:
  name: "ns-disable-mtls-pa"
  namespace: "<namespace>"
spec:
  mtls:
    mode: DISABLE
---
apiVersion: "networking.istio.io/v1alpha3"
kind: "DestinationRule"
metadata:
  name: "ns-disable-mtls-dr"
spec:
  host: "*.<namespace>.svc.cluster.local"
  trafficPolicy:
    tls:
      mode: DISABLE
```

2. Prefix protocol name to peer port or in Kubernetes 1.18+ by the appProtocol 
3. Add port 2379 to PD peer service to create listeners for pod DNS with this port as we have noticed that other components of TiDB communicate to a particular replica of PD using pod DNS and on this port

This PR addresses 2nd and 3rd only, 1st one can be manually applied for now

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manually tested on Istio 1.11.4 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

```release-note
Support for TiDB to run on Istio enabled kubernetes cluster
```

### CODE REVIEW

**Comment by DanielZhangQD** [comment]
@rahilsh Thanks for submitting the improvement！
No need to update the peer service for TiKV and TiDB?

**Comment by mikechengwei** [comment]
Can u provide an example  and readme?

**Comment by DanielZhangQD** [comment]
> Can u provide an example and readme?

@rahilsh As @mikechengwei has requested, could you please help add the example yamls required to run with Istio in the `examples` dir?

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Comment by DanielZhangQD** [comment]
> > Can u provide an example and readme?
> 
> @rahilsh As @mikechengwei has requested, could you please help add the example yamls required to run with Istio in the `examples` dir?

@rahilsh Are you available to provide a full example including the service entries if required so that the other users can refer to this to operate their own clusters? Thanks!

**Comment by rahilsh** [comment]
Sorry for late reply. Below are the steps

1. Disable MTLS by applying below yaml
```
---
apiVersion: "security.istio.io/v1beta1"
kind: "PeerAuthentication"
metadata:
  name: "ns-disable-mtls-pa"
  namespace: "<namespace>"
spec:
  mtls:
    mode: DISABLE
---
apiVersion: "networking.istio.io/v1alpha3"
kind: "DestinationRule"
metadata:
  name: "ns-disable-mtls-dr"
spec:
  host: "*.<namespace>.svc.cluster.local"
  trafficPolicy:
    tls:
      mode: DISABLE
```

2. Deploy TiDB Cluster on Kubernetes, refer https://docs.pingcap.com/tidb-in-kubernetes/stable/get-started
3. Rename port 2380 and add port 2379 to pd-peer service using below command
```
kubectl edit service <cluster_name>-pd-peer -n <namespace>
```
and then replace below in ports field
```
  ports:
  - name: tcp-peer
    port: 2380
    protocol: TCP
    targetPort: 2380
  - name: tcp-peer2
    port: 2379
    protocol: TCP
    targetPort: 2379
```

**Comment by rahilsh** [comment]
@DanielZhangQD There is no change in operator or cluster yamls, only new yaml is for disabling mtls

**Comment by rahilsh** [comment]
@DanielZhangQD Added example

**Comment by DanielZhangQD** [comment]
> @DanielZhangQD There is no change in operator or cluster yamls, only new yaml is for disabling mtls

@rahilsh Thanks a lot!
And do we have to create any service entries or any other Istio custom resources to integrate TiDB with Istio?
If yes, could you please add them to the examples?

**Comment by rahilsh** [comment]
@DanielZhangQD No other change is needed

**Comment by rahilsh** [comment]
@mikechengwei Please test it out once and let me know if you are facing any issue

File: examples/istio/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  standard (default)   kubernetes.io/gce-pd      1d
+  gold                 kubernetes.io/gce-pd      1d
+  ```
+  
+  Alternatively, you could specify the storageClass explicitly by modifying `tidb-cluster.yaml`.
+
+## Disable MTLS
+
+```bash
+> kubectl -n <namespace> apply -f disable-mtls.yaml
```
Comment (csuzhangxc) [inline examples/istio/README.md:32]: How about adding a comment to replace `<namespace>` in `disable-mtls.yaml`?

**Review by csuzhangxc — COMMENTED** [review]
rest LGTM

File: examples/istio/README.md (hunk 1 above)
Comment (rahilsh) [reply examples/istio/README.md:34]: @csuzhangxc Done

File: examples/istio/disable-mtls.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+---
+apiVersion: "security.istio.io/v1beta1"
+kind: "PeerAuthentication"
+metadata:
+  name: "ns-disable-mtls-pa"
+  namespace: "<namespace>"
```
Comment (DanielZhangQD) [inline examples/istio/disable-mtls.yaml:6]:
```suggestion
```

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**